Cookbook/Proposals/Danger-Integration.md

```diff
@@ -0,0 +1,152 @@
+
+⏱ Danger is quite fast, as all the information the rules are based on are the metadata fetched from the Pull Request or from the git diff, so it doesn't have to analyse your codebase (like a linter would do) – just to decode the JSON of the PR, run the `Dangerfile`, then call the GitHub API to post the PR comment. It's unlikely that integrating Danger will have any impact on the CI total time, especially if run in parallel of the `test_pr` jobs
+
+
+### Initial Rule to activate
+
+_As this doesn't make sense to integrate Danger in the repo without having any rule at all in the `Dangerfile` (as this would do nothing and thus would not solve any issue listed in the Motivation section), this proposal also suggest to implement one simple but useful rule during this initial Danger integration in our repo._
+
+The first rule that we propose to activate, and one which can be easy to write, is one to warn about a Pull Request which exceeds the maximum number of changes
+
+This simply consists of this code in the `Dangerfile`:
+
+```ruby
+warn("This PR is too big. Please split it in smaller PRs. If you can't split it, please justify why.") if git.lines_of_code > 800
+```
+
+This will make the bot generate a comment on the Pull Request with a ⚠️ emoji and the message above, only if the total number of lines added+removed in the diff is greater that our limit of 800 lines.
+
+Being just a warning message, it will only be informative, and will not prevent the PR to build nor make the CI fail. This is to avoid being blocked if there is a justification in breaking that limit (e.g. big refactoring that can't be split)
+
+### Additional rules in the future
+
+We already have [some ideas for additional rules](#picking-a-different-first-rule) to add to Danger that would solve some of the other issues mentioned in the [Motivation](#motivation) section.
+
+Those rules are not part of this current proposal, in order to focus discussions on integration and initial rule. We will propose those potential future rules via separate proposals.
+
+## Impact on existing codebase
+
+This tool won't have any impact on the codebase at all.
+
+It will only have an impact on our review process: we will now be informed by Danger when a PR is bigger than normal, and will be able to decide to reject that PR for being too big (or accept that it's an acceptable exception if that's justified)
+
+## Alternatives considered
+
+### Integration in the CI
+
+We could potentially trigger the `danger` CLI earlier in the CI workflow, to be notified about anything we missed ASAP in the PR. But this brings the risk of not having access to the code (if we do that before the checkout step), limiting our possibilties at what rule we can write in the future (and to be honest I'm not even sure every feature of Danger would work if we haven't checked out the code yet)
```

Comment:
I don't think there will be any issue with not having a source code when running danger, but then we will need to access the Dengerfile somehow, probably via GitHub API, which is doable, but may not worth it.

I'd suggest though to think about an option to run danger not in parallel to tests jobs, but right after it is installed (after `save_gems_cache` step in `checkout_code_steps`) so that we don't wait for things like cocoapods cache to be restored (~4-5 minutes) and persisting workspace (~2 minutes), so that the feedback will be provided after ~3 minutes (which is still long) instead of ~10

Reply:
* [ ] Add alternative to run `danger` after `save_gems_cache`

@ilyapuchka Note that if we do that, that means that `danger` will be triggered by any job that uses the `checkout_code_steps`, not limiting it to `test_pr` and `test_release` then, right?

Reply:
right, but nothing that we can't change in the config - the step can have a condition depending on the job name like we have it for steps like running `brew install`